api/src/main/java/org/apache/iceberg/transforms/Bucket.java

```diff
@@ -79,8 +77,16 @@ public Integer numBuckets() {
     return numBuckets;
   }
 
-  @VisibleForTesting
-  abstract int hash(T value);
+  @Override
+  @SuppressWarnings("unchecked")
+  public Function<T, Integer> bind(Type type) {
+    return (Function<T, Integer>) get(type, numBuckets);
+  }
+
+  protected int hash(T value) {
```

Comment:
why do we change this from an abstract method to a method with default impl that throws an exceptoin?

Reply:
I think it is needed to make `Bucket` non-abstract so that we can have unbound bucket transform.

Reply:
That's right. Bucket is no longer abstract because it doesn't need a concrete `apply` implementation. In the future, `apply` will be removed and `bind` will be used for everything.